File: crates/scriptum-lexer/src/tokens.rs
use serde::{Deserialize, Serialize};

use scriptum_utils::Span;

/// Tokens reconhecidos pelo lexer de Scriptum.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct Token {
    pub kind: TokenKind,
    pub span: Span,
}

impl Token {
    pub fn new(kind: TokenKind, span: Span) -> Self {
        Self { kind, span }
    }
}

/// Tipos de token.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum TokenKind {
    Identifier,
    NumeroLiteral,
    TextoLiteral,
    Keyword(Keyword),
    Operator(Operator),
    Delimiter(Delimiter),
    Punctuation(Punctuation),
    EOF,
}

/// Palavras-chave reservadas (latim!).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Keyword {
    Mutabilis,
    Constans,
    Functio,
    Structura,
    Si,
    Aliter,
    Dum,
    Pro,
    In,
    De,
    Redde,
    Frange,
    Perge,
    Verum,
    Falsum,
    Nullum,
    Indefinitum,
    Numerus,
    Textus,
    Booleanum,
    Vacuum,
    Quodlibet,
}

/// Delimitadores balanceados.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Delimiter {
    LParen,
    RParen,
    LBrace,
    RBrace,
    LBracket,
    RBracket,
}

/// Pontuação simples.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Punctuation {
    Colon,
    DoubleColon,
    Comma,
    Semicolon,
    Dot,
    Arrow,
    FatArrow,
    Question,
}

/// Operadores compostos.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Operator {
    Assign,
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    Pow,
    BitAnd,
    BitOr,
    BitXor,
    ShiftLeft,
    ShiftRight,
    Equal,
    StrictEqual,
    NotEqual,
    StrictNotEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,
    NullishCoalesce,
    AndAnd,
    OrOr,
    QuestionDot,
    Not,
}

File: scriptum/src/scriptum/tokens.py
"""
Lexical token definitions for the Scriptum compiler.

This module centralises the canonical list of keywords, operators, punctuation
and token kinds required by the lexer and parser.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto
from typing import Any, Dict, Iterable, Optional, Set, Tuple

from .text import Span


class TokenKind(Enum):
    """Kinds of tokens produced by the Scriptum lexer."""

    IDENTIFIER = auto()
    KEYWORD = auto()
    NUMBER_LITERAL = auto()
    STRING_LITERAL = auto()
    OPERATOR = auto()
    PUNCTUATION = auto()
    DELIMITER = auto()
    COMMENT = auto()
    WHITESPACE = auto()
    EOF = auto()


KEYWORDS: Tuple[str, ...] = (
    "mutabilis",
    "constans",
    "functio",
    "structura",
    "si",
    "aliter",
    "dum",
    "pro",
    "in",
    "de",
    "redde",
    "frange",
    "perge",
    "verum",
    "falsum",
    "nullum",
    "indefinitum",
    "numerus",
    "textus",
    "booleanum",
    "vacuum",
    "quodlibet",
)


OPERATORS: Tuple[str, ...] = (
    "=",
    "?:",
    "??",
    "||",
    "&&",
    "===",
    "!==",
    "==",
    "!=",
    ">",
    ">=",
    "<",
    "<=",
    "+",
    "-",
    "*",
    "/",
    "%",
    "**",
    "!",
    ".",
)


PUNCTUATION: Tuple[str, ...] = (",", ";", ":", "::", "->", "=>", "?")


DELIMITERS: Tuple[str, ...] = ("{", "}", "[", "]", "(", ")")


@dataclass(slots=True)
class Token:
    """Concrete token emitted by the lexer."""

    kind: TokenKind
    lexeme: str
    span: Span
    value: Optional[Any] = None
    metadata: Optional[Dict[str, Any]] = None

    def __repr__(self) -> str:  # pragma: no cover - human readable
        return f"Token(kind={self.kind.name}, lexeme={self.lexeme!r}, span={self.span!r})"


def is_keyword(lexeme: str) -> bool:
    """Check if *lexeme* is a reserved Scriptum keyword."""

    return lexeme in KEYWORDS


def all_literals() -> Iterable[str]:
    """Return the set of fixed literal lexemes (operators + punctuation + delimiters)."""

    seen: Set[str] = set()
    for collection in (OPERATORS, PUNCTUATION, DELIMITERS):
        for literal in collection:
            if literal not in seen:
                seen.add(literal)
                yield literal
